File: src/csharp/Cosmos.cs
using LanguageExt;
using LanguageExt.Common;
using LanguageExt.UnsafeValueAccess;
using Microsoft.Azure.Cosmos;
using OpenTelemetry.Trace;
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Net;
using System.Runtime.CompilerServices;
using System.Text.Json.Nodes;
using System.Threading;
using System.Threading.Tasks;

namespace common;

public sealed record ETag
{
    private readonly string value;

    private ETag(string value)
    {
        this.value = value;
    }

    public static Fin<ETag> FromString(string value) =>
        string.IsNullOrWhiteSpace(value)
            ? Error.New("ETag cannot be null or whitespace.")
            : new ETag(value);

    public static ETag Generate() => new($"\"{Guid.NewGuid()}\"");

    public static ETag All => new("\"*\"");

    public static string ToString(ETag etag) => etag.value;

    public override string ToString() => value;
}

public sealed record ContinuationToken
{
    private readonly string value;

    private ContinuationToken(string value)
    {
        this.value = value;
    }

    public static Fin<ContinuationToken> FromString(string value) =>
        string.IsNullOrWhiteSpace(value)
            ? Error.New("Continuation token cannot be null or whitespace.")
            : new ContinuationToken(value);

    public static string ToString(ContinuationToken token) => token.value;
}

public abstract record CosmosError
{
    private CosmosError() { }

    public sealed record AlreadyExists : CosmosError
    {
        private AlreadyExists() { }

        public static AlreadyExists Instance { get; } = new();
    }

    public sealed record NotFound : CosmosError
    {
        private NotFound() { }

        public static NotFound Instance { get; } = new();
    }

    public sealed record ETagMismatch : CosmosError
    {
        private ETagMismatch() { }

        public static ETagMismatch Instance { get; } = new();
    }
}

public sealed record CosmosId
{
    private readonly string value;

    private CosmosId(string value)
    {
        this.value = value;
    }

    public static Fin<CosmosId> FromString(string value) =>
        string.IsNullOrWhiteSpace(value)
            ? Error.New("Cosmos ID cannot be null or whitespace.")
            : new CosmosId(value);

    public static CosmosId Generate() => new(Guid.CreateVersion7().ToString());

    public static string ToString(CosmosId id) => id.value;

    public override string ToString() => value;
}

public record CosmosRecord<T>
{
    public required CosmosId Id { get; init; }
    public required ETag ETag { get; init; }
    public required PartitionKey PartitionKey { get; init; }
    public required T Record { get; init; }
}

public record CosmosQueryOptions
{
    public required QueryDefinition Query { get; init; }
    public Option<ContinuationToken> ContinuationToken { get; init; } = Option<ContinuationToken>.None;
    public Option<PartitionKey> PartitionKey { get; init; } = Option<PartitionKey>.None;

    public static CosmosQueryOptions FromQueryString(string query) =>
        new()
        {
            Query = new QueryDefinition(query)
        };

    public CosmosQueryOptions SetQueryString(string query) =>
        this with { Query = new QueryDefinition(query) };

    public CosmosQueryOptions SetQueryParameter(string parameterName, object parameterValue) =>
        this with { Query = Query.WithParameter(parameterName, parameterValue) };
}

public static class Cosmos
{
    public static JsonResult<CosmosId> GetId(JsonObject json) =>
        json.GetStringProperty("id")
            .Bind(id => JsonResult.FromFin(CosmosId.FromString(id)));

    public static JsonResult<ETag> GetETag(JsonObject json) =>
        json.GetStringProperty("_etag")
            .Bind(etag => JsonResult.FromFin(ETag.FromString(etag)));

    private static JsonResult<ImmutableArray<JsonObject>> GetDocumentsFromResponseJson(JsonNode json) =>
        (JsonResult<ImmutableArray<JsonObject>>)
        (from jsonObject in json.AsJsonObject()
         from jsonArray in jsonObject.GetJsonArrayProperty("Documents")
         from jsonObjects in jsonArray.GetJsonObjects()
         select jsonObjects);

    private static async Task<JsonResult<ImmutableArray<JsonObject>>> GetDocumentsFromResponse(ResponseMessage response, CancellationToken cancellationToken)
    {
        var result = await JsonNodeModule.From(response.Content, cancellationToken: cancellationToken);
        return result.Bind(GetDocumentsFromResponseJson);
    }

    private static async Task<(IEnumerable<JsonObject> Documents, Option<ContinuationToken> Token)> GetCurrentPageResults(FeedIterator iterator, CancellationToken cancellationToken)
    {
        if (!iterator.HasMoreResults)
            return (Enumerable.Empty<JsonObject>(), Option<ContinuationToken>.None);

        using var response = await iterator.ReadNextAsync(cancellationToken);
        response.EnsureSuccessStatusCode();

        var documents = await GetDocumentsFromResponse(response, cancellationToken);

        Option<ContinuationToken> continuationToken = Option<ContinuationToken>.None;
        if (!string.IsNullOrEmpty(response.ContinuationToken))
        {
            continuationToken = ContinuationToken.FromString(response.ContinuationToken)
                .Match(token => Option<ContinuationToken>.Some(token), _ => Option<ContinuationToken>.None);
        }

        return (documents.ThrowIfFail(), continuationToken);
    }

    private static async IAsyncEnumerable<JsonObject> GetFeedIteratorResults(FeedIterator iterator, [EnumeratorCancellation] CancellationToken cancellationToken)
    {
        Option<ContinuationToken> continuationToken;

        do
        {
            (var documents, continuationToken) = await GetCurrentPageResults(iterator, cancellationToken);

            foreach (var document in documents)
            {
                yield return document;
            }
        }
        while (continuationToken.IsSome);
    }

    private static FeedIterator GetFeedIterator(Container container, CosmosQueryOptions options)
    {
        var continuationToken = options.ContinuationToken
                                       .Select(ContinuationToken.ToString)
                                       .ValueUnsafe();

        var queryRequestOptions = new QueryRequestOptions();
        options.PartitionKey.Iter(partitionKey => queryRequestOptions.PartitionKey = partitionKey);

        return container.GetItemQueryStreamIterator(options.Query, continuationToken, queryRequestOptions);
    }

    public static IAsyncEnumerable<JsonObject> GetQueryResults(Container container, CosmosQueryOptions query, CancellationToken cancellationToken)
    {
        var iterator = GetFeedIterator(container, query);
        return GetFeedIteratorResults(iterator, cancellationToken);
    }

    public static async Task<Either<CosmosError, JsonObject>> ReadRecord(Container container, PartitionKey partitionKey, CosmosId id, CancellationToken cancellationToken)
    {
        using var response = await container.ReadItemStreamAsync(id.ToString(), partitionKey, cancellationToken: cancellationToken);

        switch (response.StatusCode)
        {
            case HttpStatusCode.NotFound:
                return CosmosError.NotFound.Instance;
            default:
                response.EnsureSuccessStatusCode();

                var result = from jsonNode in await JsonNodeModule.From(response.Content, cancellationToken: cancellationToken)
                             from jsonObject in jsonNode.AsJsonObject()
                             select jsonObject;

                return result.ThrowIfFail();
        }
    }

    public static async Task<Either<CosmosError, Unit>> CreateRecord<T>(Container container, PartitionKey partitionKey, T record, CancellationToken cancellationToken)
    {
        using var stream = BinaryData.FromObjectAsJson(record).ToStream();
        var options = new ItemRequestOptions { IfNoneMatchEtag = "*" };
        using var response = await container.CreateItemStreamAsync(stream, partitionKey, options, cancellationToken);

        switch (response.StatusCode)
        {
            case HttpStatusCode.Conflict:
                return CosmosError.AlreadyExists.Instance;
            default:
                response.EnsureSuccessStatusCode();

                return Unit.Default;
        }
    }

    public static async Task<Either<CosmosError, Unit>> PatchRecord(Container container, PartitionKey partitionKey, CosmosId id, ETag eTag, IEnumerable<PatchOperation> patchOperations, CancellationToken cancellationToken)
    {
        var options = new PatchItemRequestOptions { IfMatchEtag = eTag.ToString() };

        using var response = await container.PatchItemStreamAsync(id.ToString(), partitionKey, [.. patchOperations], options, cancellationToken);

        switch (response.StatusCode)
        {
            case HttpStatusCode.PreconditionFailed:
                return CosmosError.ETagMismatch.Instance;
            default:
                response.EnsureSuccessStatusCode();

                return Unit.Default;
        }
    }

    public static Task<Either<CosmosError, Unit>> SetRecordProperty(Container container, PartitionKey partitionKey, CosmosId id, ETag eTag, string propertyName, object value, CancellationToken cancellationToken) =>
        PatchRecord(container, partitionKey, id, eTag, [PatchOperation.Set($"/{propertyName}", value)], cancellationToken);

    public static Task<Either<CosmosError, Unit>> RemoveRecordProperty(Container container, PartitionKey partitionKey, CosmosId id, ETag eTag, string propertyName, CancellationToken cancellationToken) =>
        PatchRecord(container, partitionKey, id, eTag, [PatchOperation.Remove($"/{propertyName}")], cancellationToken);

    public static async Task<Either<CosmosError, Unit>> DeleteRecord(Container container, PartitionKey partitionKey, CosmosId id, ETag eTag, CancellationToken cancellationToken)
    {
        var options = new ItemRequestOptions { IfMatchEtag = eTag.ToString() };

        using var response = await container.DeleteItemStreamAsync(id.ToString(), partitionKey, options, cancellationToken);

        switch (response.StatusCode)
        {
            case HttpStatusCode.PreconditionFailed:
                return CosmosError.ETagMismatch.Instance;
            default:
                response.EnsureSuccessStatusCode();

                return Unit.Default;
        }
    }

    public static TracerProviderBuilder ConfigureCosmosOpenTelemetryTracing(this TracerProviderBuilder tracing)
    {
        AppContext.SetSwitch("Azure.Experimental.EnableActivitySource", true);

        return tracing.AddSource("Azure.Cosmos.Operation");
    }
}

File: src/fsharp/Cosmos.fs
﻿namespace common

open System
open System.Net
open Microsoft.Azure.Cosmos
open OpenTelemetry.Trace
open FSharpPlus
open FSharp.Control

type ETag =
    private
    | ETag of string

    static member fromString value =
        match String.IsNullOrWhiteSpace value with
        | true -> Error "ETag cannot be null or whitespace."
        | false -> ETag value |> Ok

    static member generate() = $"\"{Guid.NewGuid()}\"" |> ETag

    static member All = ETag "\"*\""

    static member toString(ETag value) = value

type ContinuationToken =
    private
    | ContinuationToken of string

    static member fromString value =
        match String.IsNullOrWhiteSpace value with
        | true -> Error "Continuation token cannot be null or whitespace."
        | false -> ContinuationToken value |> Ok

    static member toString(ContinuationToken value) = value

type CosmosError =
    | AlreadyExists
    | NotFound
    | ETagMismatch

type CosmosId =
    private
    | CosmosId of string

    static member fromString value =
        match String.IsNullOrWhiteSpace value with
        | true -> Error "Cosmos ID cannot be null or whitespace."
        | false -> CosmosId value |> Ok

    static member generate() =
        Guid.CreateVersion7().ToString() |> CosmosId

    static member toString(CosmosId id) = id

type CosmosRecord<'a> =
    { Id: CosmosId
      ETag: ETag
      PartitionKey: PartitionKey
      Record: 'a }

type CosmosQueryOptions =
    { Query: QueryDefinition
      ContinuationToken: ContinuationToken option
      PartitionKey: PartitionKey option }

[<RequireQualifiedAccess>]
module CosmosQueryOptions =
    let fromQueryString query =
        { CosmosQueryOptions.Query = QueryDefinition(query)
          ContinuationToken = None
          PartitionKey = None }

    let setQueryString query options =
        { options with
            Query = QueryDefinition(query) }

    let setQueryParameter parameterName parameterValue options =
        { options with
            Query = options.Query.WithParameter(parameterName, parameterValue) }

[<RequireQualifiedAccess>]
module Cosmos =
    let getId json =
        monad {
            let! idString = JsonObject.getStringProperty "id" json

            match CosmosId.fromString idString with
            | Ok cosmosId -> return cosmosId
            | Error error -> return! JsonResult.failWithMessage error
        }

    let getETag json =
        monad {
            let! etagString = JsonObject.getStringProperty "_etag" json

            match ETag.fromString etagString with
            | Ok etag -> return etag
            | Error error -> return! JsonResult.failWithMessage error
        }

    let private getDocumentsFromResponseJson json =
        json
        |> bind JsonNode.asJsonObject
        |> bind (JsonObject.getJsonArrayProperty "Documents")
        |> bind JsonArray.asJsonObjects

    let private getDocumentsFromResponse (response: ResponseMessage) =
        async {
            let! node = JsonNode.fromStream response.Content

            return getDocumentsFromResponseJson node
        }

    let private getCurrentPageResults (iterator: FeedIterator) =
        async {
            match iterator.HasMoreResults with
            | true ->
                let! cancellationToken = Async.CancellationToken
                use! response = iterator.ReadNextAsync(cancellationToken) |> Async.AwaitTask
                let _ = response.EnsureSuccessStatusCode()

                let! documents = getDocumentsFromResponse response |> Async.map JsonResult.throwIfFail

                let continuationToken =
                    ContinuationToken.fromString response.ContinuationToken |> Result.toOption

                return (documents, continuationToken)
            | false -> return (Seq.empty, None)
        }

    let private getFeedIteratorResults (iterator: FeedIterator) =
        let generator (iterator: FeedIterator) =
            match iterator.HasMoreResults with
            | true -> getCurrentPageResults iterator |> map fst |> map (fun x -> Some(x, iterator))
            | false -> async.Return None

        iterator |> AsyncSeq.unfoldAsync generator |> AsyncSeq.concatSeq

    let private getFeedIterator (container: Container) (options: CosmosQueryOptions) =
        let queryDefinition = options.Query

        let continuationToken =
            options.ContinuationToken
            |> Option.map ContinuationToken.toString
            |> Option.toObj

        let queryRequestOptions =
            let requestOptions = QueryRequestOptions()

            options.PartitionKey
            |> Option.iter (fun partitionKey -> requestOptions.PartitionKey <- partitionKey)

            requestOptions

        container.GetItemQueryStreamIterator(queryDefinition, continuationToken, queryRequestOptions)

    let getQueryResults container query =
        getFeedIterator container query |> getFeedIteratorResults

    let readRecord (container: Container) partitionKey id =
        async {
            let! cancellationToken = Async.CancellationToken
            let id = CosmosId.toString id

            use! response =
                container.ReadItemStreamAsync(id, partitionKey, cancellationToken = cancellationToken)
                |> Async.AwaitTask

            match response.StatusCode with
            | HttpStatusCode.NotFound -> return Error CosmosError.NotFound
            | _ ->
                let _ = response.EnsureSuccessStatusCode()
                let! result = JsonObject.fromStream response.Content
                return JsonResult.throwIfFail result |> Ok
        }

    let createRecord (container: Container) partitionKey record =
        async {
            use stream = BinaryData.FromObjectAsJson(record).ToStream()

            let options =
                let options = ItemRequestOptions()
                options.IfNoneMatchEtag <- "*"
                options

            let! cancellationToken = Async.CancellationToken

            use! response =
                container.CreateItemStreamAsync(stream, partitionKey, options, cancellationToken)
                |> Async.AwaitTask

            match response.StatusCode with
            | HttpStatusCode.Conflict -> return Error CosmosError.AlreadyExists
            | _ ->
                let _ = response.EnsureSuccessStatusCode()
                return Ok()
        }

    let patchRecord (container: Container) partitionKey id eTag patchOperations =
        async {
            let! cancellationToken = Async.CancellationToken

            let id = CosmosId.toString id
            let operations = List.ofSeq patchOperations |> List.map (fun operation -> operation)

            let options =
                let options = PatchItemRequestOptions()
                options.IfMatchEtag <- ETag.toString eTag
                options

            use! response =
                container.PatchItemStreamAsync(id, partitionKey, operations, options, cancellationToken)
                |> Async.AwaitTask

            match response.StatusCode with
            | HttpStatusCode.PreconditionFailed -> return Error CosmosError.ETagMismatch
            | _ ->
                let _ = response.EnsureSuccessStatusCode()
                return Ok()
        }

    let setRecordProperty (container: Container) partitionKey id eTag propertyName value =
        patchRecord container partitionKey id eTag [ PatchOperation.Set($"/{propertyName}", value) ]

    let removeRecordProperty (container: Container) partitionKey id eTag propertyName =
        patchRecord container partitionKey id eTag [ PatchOperation.Remove($"/{propertyName}") ]

    let deleteRecord (container: Container) partitionKey id eTag =
        async {
            let! cancellationToken = Async.CancellationToken

            let id = CosmosId.toString id

            let options =
                let options = ItemRequestOptions()
                options.IfMatchEtag <- ETag.toString eTag
                options

            use! response =
                container.DeleteItemStreamAsync(id, partitionKey, options, cancellationToken)
                |> Async.AwaitTask


            match response.StatusCode with
            | HttpStatusCode.PreconditionFailed -> return Error CosmosError.ETagMismatch
            | _ ->
                let _ = response.EnsureSuccessStatusCode()
                return Ok()
        }

    let configureOpenTelemetryTracing (tracing: TracerProviderBuilder) =
        AppContext.SetSwitch("Azure.Experimental.EnableActivitySource", true)

        tracing.AddSource("Azure.Cosmos.Operation")
